File: src/TrackBackupApp/AntiSpamList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using QSP.LibraryExtension;
using System.Threading;
using QSP.LibraryExtension.Tasks;
using TaskUtil = QSP.LibraryExtension.Tasks.Util;
using System.Threading.Tasks;

namespace TrackBackupApp
{
    public class AntiSpamList
    {
        public const int DefaultTokenCount = 120;
        public const int TokensPerHour = 120;

        // Maps ip to token count.
        private Dictionary<string, int> users = new Dictionary<string, int>();

        public AntiSpamList() { }

        // @NoThrow
        public async void Start()
        {
            await TaskUtil.RunPeriodic(IncrementToken,
                new TimeSpan(0, 0, 60 * 60 / TokensPerHour), new CancellationToken());
        }

        // @NoThrow
        private void IncrementToken()
        {
            var u = users;
            u.Keys.ForEach(ip =>
            {
                if (u[ip]++ >= DefaultTokenCount) users.Remove(ip);
            });

        }

        // Returns whether the ip is spammer.
        public bool DecrementToken(string ip)
        {
            var u = users;

            if (!u.TryGetValue(ip, out var count)) u[ip] = DefaultTokenCount;

            if (count == 0) return true;

            u[ip]--;
            return false;
        }

        // @NoThrow
        public bool IsSpammer(string ip)
        {
            if (!users.TryGetValue(ip, out var token)) return false;
            return token == 0;
        }
    }
}

File: src/Server_node/anti-spam-list.js
const util = require('./util')
const defaultTokenCount = 120
const tokensPerHour = 120

class AntiSpamList {
    constructor() {
        this.users = {}
        util.repeat(() => this._incrementToken(), Math.floor(60 * 60 * 1000 / tokensPerHour))
    }

    _incrementToken() {
        let u = this.users
        Object.keys(u).forEach(k => {
            if (u[k]++ >= defaultTokenCount) {
                delete u[k]
            }
        })
    }

    /**
     * @param {string} ip 
     * @returns {boolean} Whether the ip is spammer.
     */
    decrementToken(ip) {
        let u = this.users
        let count = u[ip]
        if (count === undefined) {
            u[ip] = defaultTokenCount
        }

        if (count === 0) {
            return true
        } else {
            u[ip]--
            return false
        }
    }

    isSpammer(ip) {
        return this.users[ip] === 0
    }
}

exports.AntiSpamList = AntiSpamList
